default/python/AI/FleetUtilsAI.py

```diff
@@ -621,3 +621,33 @@ def get_fleet_system(fleet):
     if isinstance(fleet, int):
         fleet = fo.getUniverse().getFleet(fleet)
     return fleet.systemID if fleet.systemID != INVALID_ID else fleet.nextSystemID
+
+
+def get_current_and_max_structure(fleet):
+    """Return a 2-tuple of the sums of structure and maxStructure meters of all ships in the fleet
+
+    :param fleet:
+    :type fleet: int | universe_object.Fleet | fo.Fleet
+    :return: tuple of sums of structure and maxStructure meters of all ships in the fleet
+    :rtype: (float, float)
+    """
+
+    universe = fo.getUniverse()
+    destroyed_ids = universe.destroyedObjectIDs(fo.empireID())
+    if isinstance(fleet, int):
+        fleet = universe.getFleet(fleet)
+    elif isinstance(fleet, Fleet):
+        fleet = fleet.get_object()
+    if not fleet:
+        return (0.0, 0.0)
+    ships_cur_health = 0
+    ships_max_health = 0
+    for ship_id in fleet.shipIDs:
+        # Check if we have see this ship get destroyed in a different fleet since the last time we saw the subject fleet
```

Comment:
What happens if we saw the ship in another fleet (but it was not destroyed yet)?

Reply:
Good point.   I think that (ages ago) when I first wrote that destruction check, there was still some bugginess with contained-object updates.  I think that's all resolved now, and so a couple lines below I've added a fleet-assignment check, which makes this destruction check *mostly* redundant (or perhaps entirely redundant).  Since the destruction check isn't dependent on any particular handling of post-destruction stats, though, I think it is a bit more reliable, plus  we still have a small bit of quirkiness with combat-granted visibility, so I'd rather still leave in this destruction check.

There are probably also a number of places where we are evaluating enemy fleets that may not be currently visible, and in any such situation a similar screening of the ships should be made, we'll have to keep our eyes out for such.